drivers/memc/memc_bflb_bl61x.c

```diff
@@ -0,0 +1,396 @@
+	tmp |= (cfg->psram_clock_divider - 1) << GLB_REG_PSRAMB_CLK_DIV_POS;
+	sys_write32(tmp, GLB_BASE + GLB_PSRAM_CFG0_OFFSET);
+
+	/* Turn on clock*/
+	tmp = sys_read32(GLB_BASE + GLB_PSRAM_CFG0_OFFSET);
+	tmp |= GLB_REG_PSRAMB_CLK_EN_MSK;
+	sys_write32(tmp, GLB_BASE + GLB_PSRAM_CFG0_OFFSET);
+}
+
+/* This initializes internal pins that are not exposed via pinctrl */
+static void memc_bflb_bl61x_init_gpio(void)
+{
+	uint32_t tmp;
+
+	for (uint8_t i = 0; i < 12; i++) {
+		tmp = GLB_REG_GPIO_41_IE_MSK | GLB_REG_GPIO_41_SMT_MSK;
+		sys_write32(tmp, GLB_BASE + GLB_GPIO_CFG41_OFFSET + i * 4);
+	}
+}
+
+static int memc_bflb_bl61x_get_psram_ctrl(const struct device *dev)
+{
+	const struct memc_bflb_bl61x_config *cfg = dev->config;
+	uint32_t tmp;
+	volatile int wait_ack = PSRAM_CONFIG_WAIT;
+
+	tmp = sys_read32(cfg->base + PSRAM_CONFIGURE_OFFSET);
+	tmp &= PSRAM_REG_CONFIG_REQ_UMSK;
+	tmp |= 1 << PSRAM_REG_CONFIG_REQ_POS;
+	sys_write32(tmp, cfg->base + PSRAM_CONFIGURE_OFFSET);
+
+	do {
+		tmp = sys_read32(cfg->base + PSRAM_CONFIGURE_OFFSET);
+		wait_ack--;
+	} while (!(tmp & PSRAM_REG_CONFIG_GNT_MSK) && wait_ack > 0);
+
+	if (wait_ack <= 0) {
+		return -ETIMEDOUT;
+	}
+	return 0;
+}
+
+static void memc_bflb_bl61x_release_psram_ctrl(const struct device *dev)
+{
+	const struct memc_bflb_bl61x_config *cfg = dev->config;
+	uint32_t tmp;
+
+	tmp = sys_read32(cfg->base + PSRAM_CONFIGURE_OFFSET);
+	tmp &= PSRAM_REG_CONFIG_REQ_UMSK;
+	sys_write32(tmp, cfg->base + PSRAM_CONFIGURE_OFFSET);
+}
+
+/* reg possible values for winbond:
+ * ID0: 0
+ * ID1: 1
+ * CR0: 2
+ * CR1: 3
+ * CR2: 4
+ * CR3: 5
+ * CR4: 6
+ */
+static int memc_bflb_bl61x_get_psram_reg(const struct device *dev, uint8_t reg, uint16_t *out)
+{
+	const struct memc_bflb_bl61x_config *cfg = dev->config;
+	uint32_t tmp;
+	volatile int wait_ack = PSRAM_CONFIG_WAIT;
+	int err;
+
+	err = memc_bflb_bl61x_get_psram_ctrl(dev);
+	if (err < 0) {
+		LOG_ERR("Get PSRAM control timed out");
+		return err;
+	}
+
+	tmp = sys_read32(cfg->base + PSRAM_CONFIGURE_OFFSET);
+	tmp &= PSRAM_REG_WB_REG_SEL_UMSK;
+	tmp |= reg << PSRAM_REG_WB_REG_SEL_POS;
+	sys_write32(tmp, cfg->base + PSRAM_CONFIGURE_OFFSET);
+
+	/* Start read from PSRAM */
+	tmp = sys_read32(cfg->base + PSRAM_CONFIGURE_OFFSET);
+	tmp &= PSRAM_REG_CONFIG_R_PUSLE_UMSK;
+	tmp |= 1 << PSRAM_REG_CONFIG_R_PUSLE_POS;
+	sys_write32(tmp, cfg->base + PSRAM_CONFIGURE_OFFSET);
+
+	do {
+		tmp = sys_read32(cfg->base + PSRAM_CONFIGURE_OFFSET);
+		wait_ack--;
+	} while (!(tmp & PSRAM_STS_CONFIG_R_DONE_MSK) && wait_ack > 0);
+
+	if (wait_ack <= 0) {
+		return -ETIMEDOUT;
+	}
+
+	tmp = sys_read32(cfg->base + PSRAM_MANUAL_CONTROL_OFFSET);
+	*out = (uint16_t)(tmp >> 16);
+
+	memc_bflb_bl61x_release_psram_ctrl(dev);
+
+	return 0;
+}
+
+static const uint16_t dqs_delay_trims[16] = {
+	0x8000,
+	0xC000,
+	0xE000,
+	0xF000,
+	0xF800,
+	0xFC00,
+	0xFE00,
+	0xFF00,
+	0xFF80,
+	0xFFC0,
+	0xFFE0,
+	0xFFF0,
+	0xFFF8,
+	0xFFFC,
+	0xFFFE,
+	0xFFFF,
+};
+
```

Comment:
I'm not sure, but coining in an idea that such tunables could be moved to the DT properties, as they describe a child hardware node the controller drives.

But given the PSRAM is embedded, it might be unnecessary

Reply:
This is a internal value for the controller related to a value set from efuse which has no root in anything available afaik so it's not very useful to set outside